gems/manageiq-providers-vmware-vcd/app/models/manageiq/providers/vmware_vcd/cloud_manager/refresher.rb

```diff
@@ -0,0 +1,71 @@
+class ManageIQ::Providers::VmwareVcd::CloudManager::Refresher < ManageIQ::Providers::BaseManager::Refresher
+  include ::EmsRefresh::Refreshers::EmsRefresherMixin
+
+  def collect_inventory_for_targets(ems, targets)
+    vcd = ems.vcd_service
+
+    inv = {
+      :orgs  => [],
+      :vdcs  => [],
+      :vapps => [],
+      :vms   => [],
+    }
+
+    vcd.organizations.all.each do |org|
+      inv[:orgs].push(org)
+    end
+
+    inv[:orgs].each do |org|
+      org.vdcs.all.each do |vdc|
+        inv[:vdcs].push(vdc)
+      end
+    end
+
+    inv[:vdcs].each do |vdc|
+      vdc.vapps.all.each do |vapp|
+        inv[:vapps].push(vapp)
+      end
+    end
+
+    inv[:vapps].each do |vapp|
+      vapp.vms.all.each do |vm|
+        inv[:vms].push(vm)
+      end
+    end
+
+    targets_with_data = targets.collect do |target|
+      _log.info "Filtering inventory for #{target.class} [#{target.name}] id: [#{target.id}]..."
+
+      case target
+      when Vm
+        # we cannot get a vm directly because it has to be loaded through vApp otherwise vapp_id is missing
+        inv[:vms].select! do |vm|
+          vm.id == target.id
+        end
+      end
+
+      _log.info "Filtering inventory...Complete"
```

Comment:
This targeted collection seems inaccurate.  Targeted refresh requires that the filtered items plus only their associations are collected.  This filters the vms, but doesn't filter anything else.